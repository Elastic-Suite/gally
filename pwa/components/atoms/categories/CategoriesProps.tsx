export interface ICategoriesPropsItem {
  isVirtual: boolean
  name: string
  path: string
  children?: ICategoriesPropsItem[]
  level: number
}
