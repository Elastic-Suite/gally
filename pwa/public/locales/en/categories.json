{
  "virtualRule.title": "Virtual rule",
  "categories.title": "Categories",
  "button.cancelSelection": "Cancel selection",
  "pinToTop": "Pin",
  "pinToBottom": "Unpin",
  "previewArea": "Preview area",
  "rows.selected_one": "{{count}} selected row.",
  "rows.selected_other": "{{count}} selected rows.",
  "allCatalogs": "All catalog",
  "allLocales": "All locals",
  "catalog.dropdown.label": "Choose a catalog",
  "localizedCatalog.dropdown.label": "Choose a locale",
  "searchBar.title": "Products in category",
  "searchBar.results_one": "result",
  "searchBar.results_zero": "result",
  "searchBar.results_other": "results",
  "searchBar.placeholder": "Search by name...",
  "required": "* Required field",
  "name": "Use category name in product search",
  "name.tooltip": "Use category name in product search",
  "virtual": "Virtual category",
  "virtual.tooltip": "Virtual category",
  "select": "Select",
  "select.tooltip": "What sorting will be used for the category",
  "searchBarResult": "{{value.value}} {{value.result}}",
  "placeholder": "No category selected"
}
