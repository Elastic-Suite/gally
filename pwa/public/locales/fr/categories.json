{
  "virtualRule.title": "Règle virtuelle",
  "categories.title": "Catégories",
  "button.cancelSelection": "Annuler la sélection",
  "pinToTop": "Pin",
  "pinToBottom": "Unpin",
  "previewArea": "Zone de prévisualisation",
  "rows.selected_one": "{{count}} ligne sélectionnée",
  "rows.selected_other": "{{count}} lignes sélectionnées",
  "allCatalogs": "Tous les catalogues",
  "allLocales": "Toutes les locales",
  "catalog.dropdown.label": "Choisir un catalogue",
  "localizedCatalog.dropdown.label": "Choisir une locale",
  "searchBar.title": "Produits dans la catégorie",
  "searchBar.results_one": "résultat",
  "searchBar.results_zero": "résultat",
  "searchBar.results_other": "résultats",
  "searchBar.placeholder": "Recherche par nom...",
  "required": "* Champ requis",
  "name": "Utiliser le nom de la catégorie dans la recherche de produits",
  "name.tooltip": "Utiliser le nom de la catégorie dans la recherche de produits",
  "virtual": "Catégorie virtuelle",
  "virtual.tooltip": "Catégorie virtuelle",
  "select": "Sélectionnez",
  "select.tooltip": "Quel tri sera utilisé pour la catégorie",
  "searchBarResult": "{{value.value}} {{value.result}}",
  "placeholder": "Aucune catégorie selectionnée."
}
