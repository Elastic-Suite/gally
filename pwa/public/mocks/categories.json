[
  {
    "id": "1",
    "name": "Categ 1",
    "level": 1,
    "path": "1",
    "isVirtual": false,
    "children": [
      {
        "id": "11",
        "name": "Categ 11",
        "level": 2,
        "path": "1/11",
        "isVirtual": true
      },
      {
        "id": "12",
        "name": "Categ 12",
        "level": 2,
        "path": "1/12",
        "isVirtual": true
      }
    ]
  },
  {
    "id": "2",
    "name": "Categ 2",
    "level": 1,
    "path": "2",
    "isVirtual": false,
    "children": [
      {
        "id": "21",
        "name": "Categ 21",
        "level": 2,
        "path": "1/21",
        "isVirtual": true,

        "children": [
          {
            "id": "211",
            "name": "Categ 211",
            "level": 3,
            "path": "1/21/211",
            "isVirtual": false,

            "children": [
              {
                "id": "2111",
                "name": "Categ 2111",
                "level": 4,
                "path": "1/21/211/2111",
                "isVirtual": true
              }
            ]
          },
          {
            "id": "212",
            "name": "Categ 212",
            "level": 3,
            "path": "1/21/212",
            "isVirtual": true
          }
        ]
      }
    ]
  },
  {
    "id": "3",
    "name": "Categ 3",
    "level": 1,
    "path": "3",
    "isVirtual": false
  }
]
