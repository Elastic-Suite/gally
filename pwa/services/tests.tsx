import { RenderOptions, render } from '@testing-library/react'
import { renderHook } from '@testing-library/react-hooks'
import { PreloadedState } from '@reduxjs/toolkit'
import { ReactChild } from 'react'

import AppProvider from '~/components/AppProvider'
import { AppStore, RootState, setupStore } from '~/store'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactChild }): JSX.Element {
    return <AppProvider store={store}>{children}</AppProvider>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactChild }): JSX.Element {
    return <AppProvider store={store}>{children}</AppProvider>
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
