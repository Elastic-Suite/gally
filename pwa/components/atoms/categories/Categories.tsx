import { useState } from 'react'
import { styled, keyframes } from '@mui/system'
import IonIcon from '../IonIcon/IonIcon'
import { ICategoriesPropsItem } from './CategoriesProps'

const CustomRoot = styled('ul')({
  margin: 0,
  padding: 0,
  fontFamily: 'Inter',
  fontWeight: 500,
})

const opacity = keyframes`
from {
  opacity: 0;
}

to {
  opacity: 1;
}
`
const CustomLi = styled('li')(({ theme }) => ({
  listStyleType: 'none',
  marginBottom: theme.spacing(1),
  marginTop: theme.spacing(1),
  opacity: 0,
  animation: `${opacity} 1000ms forwards`,
  '&:hover': {
    cursor: 'pointer',
  },
}))

const CustomTitle = styled('div')(({ theme }) => ({
  color: theme.palette.colors.secondary[600],
  position: 'relative',
  '&:hover::before': {
    content: '""',
    position: 'absolute',
    width: '100%',
    height: '1px',
    top: '100%',
    fontSize: '12px',
    background: theme.palette.colors.secondary[600],
  },
}))

const CustomTitleBase = styled('div')(({ theme }) => ({
  color: theme.palette.colors.neutral[900],
  position: 'relative',
  '&:hover::before': {
    content: '""',
    position: 'absolute',
    width: '100%',
    height: '1px',
    top: '100%',
    fontSize: '12px',
    background: theme.palette.colors.neutral[900],
  },
}))

const CustomVirtual = styled('span')(({ theme }) => ({
  color: theme.palette.colors.neutral['900'],
  fontSize: '10px',
  fontWeight: 600,
  lineHeight: '12px',
  background: theme.palette.colors.neutral['300'],
  paddingTop: '2px',
  paddingBottom: '2px',
  paddingRight: '4px',
  paddingLeft: '4px',
  borderRadius: theme.spacing(1),
  marginLeft: theme.spacing(0.75),
}))

const CustomBtn = styled('span')(({ theme }) => ({
  color: theme.palette.colors.neutral['500'],
  borderRadius: '11px',
  width: '28px',
  boxSizing: 'border-box',
  height: '28px',
  fontSize: '28px',
  transition: 'transform 100ms linear',
  display: 'flex',
  justifyContent: 'center',
  padding: '2px',
  alignItems: 'center',
  '&:hover': {
    transform: 'rotate(180deg)',
    color: theme.palette.colors.secondary['600'],
    background: theme.palette.colors.secondary['100'],
  },
}))

const CustomLink = styled('a')({
  textDecoration: 'none',
})

interface IProps {
  data: ICategoriesPropsItem[]
}

const Categories = ({ data }: IProps) => {
  const [displayChildren, setDisplayChildren] = useState({})

  return (
    <CustomRoot>
      {data.map((item, key) => {
        const Title = item.level === 1 ? CustomTitleBase : CustomTitle
        return (
          <CustomLi style={{ marginLeft: item.level === 1 ? 0 : 30 }} key={key}>
            <div style={{ display: 'flex', gap: '2px', alignItems: 'center' }}>
              {item.children && (
                <CustomBtn
                  onClick={() => {
                    setDisplayChildren({
                      ...displayChildren,
                      [item.name]: !displayChildren[item.name],
                    })
                  }}
                >
                  {displayChildren[item.name] ? (
                    <IonIcon name="minus" />
                  ) : (
                    <IonIcon name="more" />
                  )}
                </CustomBtn>
              )}
              <div
                style={{
                  display: 'flex',
                  alignItems: 'center',
                  justifyContent: 'center',
                }}
              >
                <Title
                  onClick={() => {
                    setDisplayChildren({
                      ...displayChildren,
                      [item.name]: !displayChildren[item.name],
                    })
                  }}
                >
                  {!item.children ? (
                    <CustomLink href={item.path}>{item.name}</CustomLink>
                  ) : (
                    item.name
                  )}
                </Title>
                {item.isVirtual && <CustomVirtual>virtual</CustomVirtual>}
              </div>
            </div>
            {displayChildren[item.name] && item.children && (
              <Categories data={item.children} />
            )}
          </CustomLi>
        )
      })}
    </CustomRoot>
  )
}

export default Categories
