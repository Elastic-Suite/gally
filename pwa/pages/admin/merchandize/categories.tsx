import { useContext, useEffect, useMemo, useState } from 'react'
import { useTranslation } from 'next-i18next'
import Head from 'next/head'
import { useRouter } from 'next/router'
import { Box } from '@mui/system'

import { emptyCombinationRule } from '~/constants'
import { breadcrumbContext } from '~/contexts'
import { withAuth } from '~/hocs'
import {
  useApiList,
  useFetchApi,
  useResource,
  useResourceOperations,
} from '~/hooks'
import { findBreadcrumbLabel } from '~/services'
import { selectMenu, useAppSelector } from '~/store'
import { ICatalog, ICategories, ICategory, ISearchParameters, IHydraResponse, } from '~/types'

import TitleBlock from '~/components/molecules/layout/TitleBlock/TitleBlock'
import TwoColsLayout from '~/components/molecules/layout/twoColsLayout/TwoColsLayout'
import CatalogSwitcher from '~/components/stateful/CatalogSwitcher/CatalogSwitcher'
import CategoryTree from '~/components/stateful/CategoryTree/CategoryTree'
import ProductsContainer, {
  IConfiguration,
} from '~/components/stateful/ProductsContainer/ProductsContainer'
import RulesManager from '~/components/stateful/RulesManager/RulesManager'

function Categories(): JSX.Element {
  const router = useRouter()
  const menu = useAppSelector(selectMenu)
  const [, setBreadcrumb] = useContext(breadcrumbContext)

  const [selectedCategoryItem, setSelectedCategoryItem] = useState<ICategory>()

  const resourceName = 'Catalog'
  const resource = useResource(resourceName)
  const [catalogsFields] = useApiList<ICatalog>(resource, false)
  const { data, error } = catalogsFields

  const [catalogId, setCatalogId] = useState<number>(-1)
  const [localizedCatalogId, setLocalizedCatalogId] = useState<number>(-1)
  const [rule, setRule] = useState(emptyCombinationRule)

  useEffect(() => {
    setBreadcrumb(['merchandize', 'categories'])
  }, [router.query, setBreadcrumb])

  const title = findBreadcrumbLabel(0, ['merchandize'], menu.hierarchy)

  const { t } = useTranslation('categories')

  const [categories] = useFetchApi<ICategories>(
    `categoryTree?/&catalogId=${
      catalogId !== -1 ? catalogId : null
    }&localizedCatalogId=${
      localizedCatalogId !== -1 ? localizedCatalogId : null
    }`
  )

  const categoryConfigurationParams: ISearchParameters = useMemo(
    () => ({
      categoryId: selectedCategoryItem ? selectedCategoryItem.id : null,
      catalogId: catalogId !== -1 ? catalogId : null,
      localizedCatalogId: localizedCatalogId !== -1 ? localizedCatalogId : null,
    }),
    [selectedCategoryItem, catalogId, localizedCatalogId]
  )
  const categoryConfigurationResource = useResource('CategoryConfiguration')
  const [dataCats, updateDataCats] = useFetchApi<
    IHydraResponse<IConfiguration>
  >(categoryConfigurationResource, categoryConfigurationParams)

  const dataCat =
    catalogId !== -1 &&
    localizedCatalogId !== -1 &&
    dataCats?.data['hydra:member']
      ? dataCats?.data['hydra:member'].find((el) => el?.id === catalogId)
      : undefined

  const idCat = dataCat?.id

  const configuration = useResource('CategoryConfiguration')
  const { update } = useResourceOperations<IConfiguration>(configuration)

  const [saveData, setSaveData] = useState({})

  function handleUpdateCat(name: string): (val: boolean | string) => void {
    return (val) => {
      if (catalogId !== -1 && localizedCatalogId !== -1) {
        setSaveData((state) => ({ ...state, [name]: val }))

        updateDataCats((categ) => {
          return {
            ...categ,
            'hydra:member': categ['hydra:member'].map((el) => {
              if (el.id === catalogId) return { ...el, [name]: val }
              return el
            }),
          }
        })
      }
    }
  }

  const [saveTest, setSaveTest] = useState({})

  useEffect(() => {
    if (
      localizedCatalogId !== -1 &&
      catalogId !== -1 &&
      !Object.entries(saveData).length
    )
      setSaveTest({
        useNameInProductSearch: dataCat.useNameInProductSearch,
        isVirtual: dataCat.isVirtual,
        defaultSorting: dataCat.defaultSorting,
      })
  }, [dataCat])

  async function onSave(): Promise<void> {
    await update(idCat, saveData)
    setSaveData({})
  }

  function testBtnSaveDisabled(): boolean {
    if (
      Object.entries(saveTest).find(
        ([key, val]: [
          key: (keyof typeof saveTest)['string'],
          val: string | boolean
        ]) => {
          if (saveData[key] === undefined) return 0
          else if (saveData[key] === val) return 0
          return 1
        }
      )
    )
      return false
    return true
  }

  const disableBtnSave = testBtnSaveDisabled()

  return (
    <>
      <Head>
        <title>{title}</title>
      </Head>
      <TwoColsLayout
        left={[
          <TitleBlock key="categories" title={t('categories.title')}>
            <>
              <CatalogSwitcher
                catalog={catalogId}
                onCatalog={setCatalogId}
                localizedCatalog={localizedCatalogId}
                onLocalizedCatalog={setLocalizedCatalogId}
                catalogsData={data}
                error={error}
                onCategory={setSelectedCategoryItem}
              />
              <CategoryTree
                categories={categories.data}
                selectedItem={selectedCategoryItem}
                onSelect={setSelectedCategoryItem}
              />
            </>
          </TitleBlock>,
          <TitleBlock key="virtualRule" title={t('virtualRule.title')}>
            <VirtualRule val={dataCat?.isVirtual}>
              <RulesManager
                catalogId={catalogId}
                localizedCatalogId={localizedCatalogId}
                onChange={setRule}
                rule={rule}
              />
            </VirtualRule>
          </TitleBlock>,
        ]}
      >
        {selectedCategoryItem?.id ? (
          <ProductsContainer
            category={selectedCategoryItem}
            dataCat={dataCat}
            onVirtualChange={handleUpdateCat('isVirtual')}
            onNameChange={handleUpdateCat('useNameInProductSearch')}
            onSortChange={handleUpdateCat('defaultSorting')}
            catalog={catalogId}
            localizedCatalog={localizedCatalogId}
            catalogsData={data}
            error={error}
            onSave={onSave}
            disableBtnSave={disableBtnSave}
          />
        ) : (
          <Box
            sx={{
              fontSize: '12px',
              fontFamily: 'inter',
              lineHeight: '18px',
              padding: '16px 0  0 16px',
              color: 'colors.neutral.600',
            }}
          >
            {t('placeholder')}
          </Box>
        )}
      </TwoColsLayout>
    </>
  )
}

export default withAuth(Categories)
