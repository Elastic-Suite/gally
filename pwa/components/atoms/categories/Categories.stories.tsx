import { ComponentMeta, ComponentStory } from '@storybook/react'
import Categories from './Categories'
import categories from '../../../public/mocks/categories.json'

export default {
  title: 'Atoms/Categories',
  component: Categories,
} as ComponentMeta<typeof Categories>

const Template: ComponentStory<typeof Categories> = (args) => (
  <Categories {...args} />
)

export const Default = Template.bind({})
Default.args = {
  items: categories,
}
