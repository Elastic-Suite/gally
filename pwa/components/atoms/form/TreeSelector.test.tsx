import { renderWithProviders } from '~/utils/tests'
import categories from '~/public/mocks/categories.json'

import TreeSelector from './TreeSelector'

describe('TreeSelector', () => {
  it('should match snapshot', () => {
    const { container } = renderWithProviders(
      <TreeSelector data={categories.categories} value={[]} />
    )
    expect(container).toMatchSnapshot()
  })
})
