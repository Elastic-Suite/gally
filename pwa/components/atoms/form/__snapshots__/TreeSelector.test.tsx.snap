// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeSelector should match snapshot 1`] = `
<div>
  <div
    class="css-1baulvz"
  >
    <div
      class="MuiFormControl-root css-14yckze-MuiFormControl-root"
    >
      <div
        class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd  css-xrz2lz-MuiInputBase-root"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-yz9k0d-MuiInputBase-input"
          id="mui-1"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1uab8qk-MuiInputAdornment-root"
        >
          <button
            aria-label="form.open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1x878dy-MuiButtonBase-root-MuiIconButton-root"
            tabindex="-1"
            title="form.open"
            type="button"
          >
            <ion-icon
              name="chevron-down"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
